music_assistant/server/providers/plex/__init__.py

```diff
@@ -0,0 +1,467 @@
+            return plex_account.resource(self.config.get_value(CONF_SERVER_NAME)).connect()
+
+        self._plex_server = await self._run_async(connect)
+        self._plex_library = await self._run_async(
+            self._plex_server.library.section, self.config.get_value(CONF_LIBRARY_NAME)
+        )
+
+    async def resolve_image(self, path: str) -> str | bytes | AsyncGenerator[bytes, None]:
+        """Return the full image URL including the auth token."""
+        return self._plex_server.url(path, True)
+
+    @property
+    def supported_features(self) -> tuple[ProviderFeature, ...]:
+        """Return a list of supported features."""
+        return (
+            ProviderFeature.LIBRARY_ARTISTS,
+            ProviderFeature.LIBRARY_ALBUMS,
+            ProviderFeature.LIBRARY_TRACKS,
+            ProviderFeature.LIBRARY_PLAYLISTS,
+            ProviderFeature.BROWSE,
+            ProviderFeature.SEARCH,
+            ProviderFeature.ARTIST_ALBUMS,
+        )
+
+    async def _run_async(self, call: Callable, *args, **kwargs):
+        return await self.mass.create_task(call, *args, **kwargs)
+
+    async def _get_data(self, key, cls=None):
+        return await self._run_async(self._plex_library.fetchItem, key, cls)
+
+    @alru_cache(maxsize=128, ttl=300)
+    async def _get_data_cached(self, key, ma_cls):
+        if ma_cls == Artist:
+            return await self._parse_artist(await self._get_data(key, PlexArtist))
+        if ma_cls == Album:
+            return await self._parse_album(await self._get_data(key, PlexAlbum))
+
+    async def _parse(self, plex_media):
+        if plex_media.type == "artist":
+            return await self._parse_artist(plex_media)
+        elif plex_media.type == "album":
+            return await self._parse_album(plex_media)
+        elif plex_media.type == "track":
+            return await self._parse_track(plex_media)
+        elif plex_media.type == "playlist":
+            return await self._parse_playlist(plex_media)
+        return None
+
+    async def _search_track(self, search_query, limit) -> list[PlexTrack]:
+        return await self._run_async(
+            self._plex_library.searchTracks, title=search_query, limit=limit
+        )
+
+    async def _search_album(self, search_query, limit) -> list[PlexAlbum]:
+        return await self._run_async(
+            self._plex_library.searchAlbums, title=search_query, limit=limit
+        )
+
+    async def _search_artist(self, search_query, limit) -> list[PlexArtist]:
+        return await self._run_async(
+            self._plex_library.searchArtists, title=search_query, limit=limit
+        )
+
+    async def _search_playlist(self, search_query, limit) -> list[PlexPlaylist]:
+        return await self._run_async(self._plex_library.playlists, title=search_query, limit=limit)
+
+    async def _search_track_advanced(self, limit, **kwargs) -> list[PlexTrack]:
```

Comment:
whats the advanced for ?

Reply:
Plex offers a really cool and complex query language, where you can build full decision trees. But for now these search methods are all not used. Does it make sense to keep them because MA supports that kind of search, too or is it more like we have a search string an a media_type?

Reply:
Cool, yes MA supports these kind of queries too (for all the sorting and stuff in the frontend) but this is done on MA's own database as abstraction layer because there is no way we can find common ground between all the different providers.

But yeah, keep it in here because maybe we find a future use ;-)